From 0af98cc1baf98d939a9dcbcb512c12b89300ef24 Mon Sep 17 00:00:00 2001
From: Frank Wang <frank.wang@rock-chips.com>
Date: Thu, 12 Mar 2020 10:17:11 +0800
Subject: [PATCH] usb: dwc2: fix otg host fail after system resume

Fix commit 079b3c7cdb59 ("usb: dwc2: make hcd into L3 power
off state when suspend").

This fixes otg-host restore failed after system resume for
dwc2 related interrupts were not enabled when no gadget had
been bounded before.

Change-Id: Iceac66acf063c6226dcafb2086149122511ab5c1
Signed-off-by: Frank Wang <frank.wang@rock-chips.com>
---
 drivers/usb/dwc2/platform.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/usb/dwc2/platform.c b/drivers/usb/dwc2/platform.c
index dfbb6459a407..d0012bce15ee 100644
--- a/drivers/usb/dwc2/platform.c
+++ b/drivers/usb/dwc2/platform.c
@@ -605,6 +605,7 @@ static int __maybe_unused dwc2_suspend(struct device *dev)
 static int __maybe_unused dwc2_resume(struct device *dev)
 {
 	struct dwc2_hsotg *dwc2 = dev_get_drvdata(dev);
+	unsigned long flags;
 	int ret = 0;
 
 	if (dwc2->ll_hw_enabled) {
@@ -615,9 +616,13 @@ static int __maybe_unused dwc2_resume(struct device *dev)
 
 	/* Stop hcd if dr_mode is host and PD is power off when suspend */
 	if (dwc2->op_state == OTG_STATE_A_HOST && dwc2_is_device_mode(dwc2)) {
+		spin_lock_irqsave(&dwc2->lock, flags);
 		dwc2_hcd_disconnect(dwc2, true);
 		dwc2->op_state = OTG_STATE_B_PERIPHERAL;
 		dwc2->lx_state = DWC2_L3;
+		if (!dwc2->driver)
+			dwc2_hsotg_core_init_disconnected(dwc2, false);
+		spin_unlock_irqrestore(&dwc2->lock, flags);
 	}
 
 	if (dwc2_is_device_mode(dwc2))
-- 
2.35.3

